### DISCUSSION TITLE
LED light flickering due to PWM

### DISCUSSION THREAD

**Discussion by xblax** [xblax/flashforge_ad5m_klipper_mod#136] [General]
I'm wondering if others also have that issue, but didn't see it mentioned anywhere. I modded my ADM5 to have 24V LEDs: I have installed two stripes, one on each side, similar to https://www.thingiverse.com/thing:6358773

When lowering the intensity, the PWM causes flickering on the camera. Does the 5M Pro build-in LED avoid this somehow or does it flicker as well? 

I can be mitigated a bit by decreasing the PWM cycle, but Klipper docs do not really recommend it.
https://www.klipper3d.org/Config_Reference.html#led

**Comment by KaruroChori** [comment]
Are you observing it in a "high" frame-rate video capture (like between 15-60 fps) or in a long time-lapse?
What is the base frequency used for the pwm signal compared to the shutter speed of your camera?

**Comment by xblax** [comment]
It's immediately visible in the live stream which as 30fs. Just like when taking videos or stills with cheap LED lighting with low PWM cycle. Due to the rolling shutter of the camera sensor, there are very visible flickering artifacts.  

Don't know what the shutter speed of the webcam is. I guess it is auto-adjusting according to brightness.

Flickering improves at certain LED brightness levels if the LED PWM matches the framerate. Higher PWM cycle gives better results.

**Reply by KaruroChori** [reply]
> Higher PWM cycle gives better results.

Yes, before you wrote `decreasing the PWM cycle` which confused me because it was not clear if you meant time or frequency.  
As long as the board has a hardware pwm channel to spare or the processor can keep up with the software pwm, increasing the frequency (or decreasing the period) should do the trick. 

Other than that, the only solution is to have a better camera with global shutter or a faster sensor readout.

The only issue is that the former are not particularly cheap :D
For example https://www.e-consystems.com/de/industrial-cameras/ar0234-usb3-global-shutter-kamera-de.asp

**Reply by xblax** [reply]
Well technically shorter PWM cycle time (decreasing) or higher PWM frequency. But I guess we are meaning the same thing 😄 
Sure, I'm aware of technical solutions like hardware PWM or better cameras. But I'm using the default LED port on the mainboard and official camera. That's why I'm wondering if that is a common issue for everyone.

Software PWM frequency can be increased quite a bit and could be tested if it causes issues. But if it's only me having the issue, I just do it for my own config.

**Reply by consp** [reply]
> Software PWM frequency can be increased quite a bit and could be tested if it causes issues.

Strange that they use software, it's connected to the pin of TIM1_CH1 so my guess is it looks like the intent was to use a hardware pwm output.

Most commercial stuff uses 5KHz for 25/30fps videos but that's a bit much for software pwm :laughing: . The default is 100Hz which is pretty much dead-on crap. If the shutter frequency is reliable you can also find a dead-on frequency and it will be either light or dark :rofl: .

**Comment by Tiwatz** [comment]
> Does the 5M Pro build-in LED avoid this somehow or does it flicker as well?
> 

I've noticed some flickering on my Pro model as well.

**Reply by consp** [reply]
Me too but I'm not bothered with it much, one of my LED bulbs also suffers from this (due to cheap capacitive droppers used in them) and it will be present anyway.

**Comment by Matt-Armfield** [comment]
I have flickering with the 24V COB LED strips I have placed in my 5M.  They are slightly noticeable when the print head is moving by eye but greatly exaggerating but the factory camera.  I've thought about putting a capacitor on the leads to the LED strip to see if that smooths it out but haven't had my printer off long enough to do this mod.

**Reply by ANKY06** [reply]
I have flickering with the 24V COB LED strips I have placed in my 5M as well. I have noticed that the flickering happens when the bed heater is not Off or not 100% On (which is when the bed has reached the defined temperature and maintaining it). I havent tried shotening the strip yet. Adding capacitors might help with the flickering but mess with the PWM.
